include/chemfiles/shared_allocator.hpp

```diff
@@ -0,0 +1,185 @@
+
+#include <unordered_map>
+#include <functional>
+#include <cassert>
+#include <vector>
+#include <atomic>
+
+#include "chemfiles/ErrorFmt.hpp"
+
+/// An atomic counter, for reference counted pointers
+class atomic_count {
+public:
+    explicit atomic_count(long value): value_(static_cast<std::int_least32_t>(value)) {}
+    ~atomic_count() = default;
+
+    atomic_count(const atomic_count&) = delete;
+    atomic_count& operator=(const atomic_count&) = delete;
+
+    atomic_count(atomic_count&& other): value_(static_cast<std::int_least32_t>(other.load())) {
+        other.value_ = 0;
+    }
+
+    atomic_count& operator=(atomic_count&& other) {
+        value_ = static_cast<std::int_least32_t>(other.load());
+        other.value_ = 0;
+        return *this;
+    }
+
+    // Increase counter by one and return the resulting value
+    long increase() {
+        return value_.fetch_add(1, std::memory_order_acq_rel) + 1;
+    }
+
+    // Decrease the counter by one and return the resulting value
+    long decrease() {
+        return value_.fetch_sub(1, std::memory_order_acq_rel) - 1;
+    }
+
+    // Get the current counter value
+    long load() const {
+        return value_.load(std::memory_order_acquire);
+    }
+
+private:
+    std::atomic_int_least32_t value_;
+};
+
+
+/// An allocator with shared_ptr like semantics, working with raw pointers.
+///
+/// This is used in the C API to ensure that when taking pointers to
+/// atoms/residues/cell inside a frame/topology, the frame/topology is keeped
+/// alive even if the user call chfl_xxx_free.
+class shared_allocator {
+public:
+    shared_allocator() = default;
+    shared_allocator(const shared_allocator&) = delete;
+    shared_allocator& operator=(const shared_allocator&) = delete;
+    shared_allocator(shared_allocator&&) = delete;
+    shared_allocator& operator=(shared_allocator&&) = delete;
+
+    /// Like std::make_shared: create a new shared pointer by constructing a
+    /// value of type T with the given arguments.
+    template<class T, typename ... Args>
+    static T* make_shared(Args&& ... args) {
+        auto ptr = new T{std::forward<Args>(args)...};
+        instance_.insert_new(ptr);
+        return ptr;
+    }
+
+    /// Like std::shared_ptr<U> aliasing contructor: element and ptr will share
+    /// the references count, and none will be freed while the other one is
+    /// alive.
+    ///
+    /// `ptr` must have been allocated with make_shared.
+    template<class T, class U>
+    static T* shared_ptr(U* ptr, T* element) {
+        instance_.insert_shared(ptr, element);
+        auto& count = instance_.count(ptr);
+        count.increase();
+        return element;
+    }
+
+    template<class T, class U>
+    static const T* shared_ptr(U* ptr, const T* element) {
+        return shared_ptr(ptr, const_cast<T*>(element));
+    }
+
+    /// Decrease the reference count of `ptr`, and delete it if needed.
+    static void free(const void* ptr) {
```

Comment:
Ok, I understand - still on the fence about freeing `const` memory from an API perspective.